src/Bonus.java

```diff
@@ -1,66 +1,89 @@
+// Opmerking vooraf: Ik zal eerlijk zijn, ik heb niet alle code volledig zelf geschreven. Een opdracht als deze is nog behoorlijk pittig, en daarom heb ik o.a. ChatGPT gebruikt bij de gedeelten waar ik niet uitkwam, en ook bij nieuwe dingen die ik leuk vond om toe te voegen (error handling en betere user feedback). Ik heb me uiteraard wel serieus in de code verdiept, en ook bij de niet zelf bedachte code gezorgd dat ik van elke regel begrijp wat er gebeurt. Dat maakte het toch ook heel leerzaam en ik vond het ontzettend leuk! Met een beetje hulp kon ik al dingen maken die ik nog niet voor mogelijk had gehouden in deze fase.
+
 import java.util.HashSet;
 import java.util.Objects;
+import java.util.Random;
 import java.util.Scanner;
 
 public class Bonus {
 
     public static void main(String[] args) {
 
-        // Onderstaande code is uitgecommentarieerd. Haal de code uit de comment (delete /* en */) om aan de slag te gaan met de bonus
-        // In de code staan op verschillende plekken comments tussen "/*...*/" als hint wat je moet doen
-
-
-        HashSet<Integer> secretnumber = randomnumbergenerator();
-        String stringnumber = setToStringConverter(secretnumber);
-        System.out.println(stringnumber);
-        feedback(/*vul hier het juiste argument in*/);
-
+        // Hier worden alle gebruikte methodes aangeroepen, die daaronder worden gedefinieerd
+        HashSet<Integer> secretNumber = randomNumberGenerator();
+        String stringNumber = setToString(secretNumber);
+        System.out.println(stringNumber);
+        feedback(stringNumber);
     }
 
-    public static void/*moet dit returntype "void" zijn of wat anders?*/ randomnumbergenerator(/*Heeft deze methode nog parameter(s) nodig?*/){
-         /*
-        Vul hier de body van de methode in.
+    public static HashSet<Integer> randomNumberGenerator() { // De return value is de HashSet<Integer>, er zijn geen parameters mee te geven.
+        Random random = new Random(); // Maakt een nieuwe instantie van de Random class
+        HashSet<Integer> secretNumber = new HashSet<>(); // Maakt een nieuwe HashSet, die in eerste instantie leeg is.
+        while (secretNumber.size() < 4) { // Zolang de set kleiner is dan 4
+            int randomNumber = random.nextInt(9) + 1; // wordt er een random number gegenereerd tussen 0 en 9 en opgeslagen als variabele
```

Comment:
Indien je hier int randomNumber = random.nextInt(10) had gebruikt, dan zou er een random number gegenereerd worden tussen 0 en 9. Echter jij hebt gekozen voor random.nextInt(9). Dat houdt in dat hij tot 8 gaat en omdat jij een + 1 op het eind hebt gedaan zal bij jou een getal tussen 1 en 9 gekozen worden, geen 0.

Reply:
Nou, dan heb je beter opgelet dan de makers van deze opdracht! In de officiële uitwerkingen hebben ze het namelijk net zo gedaan als ik. Maar het klopt dan je dan inderdaad geen 0 kan kiezen.